### ISSUE TITLE
PythonLoggingObserver does not log structured data.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5696]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)| @dreid reported|
|-|-|
|Trac ID|trac#5696|
|Type|defect|
|Created|2012-05-30 22:14:52Z|
|Branch|https://github.com/twisted/twisted/tree/structured-log-logging-5696|

The stdlib logging module supports an `extra` keyword argument which is a dictionary that will be combined with the `LogRecord`.  When `PythonLoggingObserver` calls `self.logging.log` it fails to pass the `eventDict` argument to `emit` as this `extra` keyword argument, thus all structured data that might be attached to the log message is lost.

The `extra` keyword argument is documented here: http://docs.python.org/library/logging.html#logging.Logger.debug

Attachments:

* [logging-extra.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a9c/a9c9a6e266a80dc17285537584f10a5d4322f5e8/1d358d34eb4255110e646736331606ff7b000c58.patch) (1208 bytes) - added by chris- on 2012-05-31 11:57:49Z - Pass eventDict along to stdlib logging as extras
* [logging-extra-v2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a9c/a9c9a6e266a80dc17285537584f10a5d4322f5e8/1cf9b1137a38f91f6af6133d71dfe2e11d3cd2bf.patch) (2624 bytes) - added by chris- on 2012-06-07 10:31:19Z - Replaces patch logging-extra.patch
* [logging-extra-v2.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a9c/a9c9a6e266a80dc17285537584f10a5d4322f5e8/0aa83a28c85c2e80ad285c0665786c191f3ae773.patch) (799 bytes) - added by richzeng on 2012-06-09 00:12:12Z - Meant to replace the original patch to twisted.python.log. Other processes make use of the eventDict, so we can't modify it in place. We need to make a copy of eventDict (in this case 'extraDict'). We also need to filter out the extraDict for certain keys that will make Python's stdlib complain when you try to pass them in with the 'extra' keyword.

<details><summary>Searchable metadata</summary>

```
trac-id__5696 5696
type__defect defect
reporter__dreid dreid
priority__normal normal
milestone__ 
branch__branches_structured_log_logging_5696 branches/structured-log-logging-5696
branch_author__glyph glyph
status__new new
resolution__None None
component__core core
keywords__ 
time__1338416092000000 1338416092000000
changetime__1429134760057938 1429134760057938
version__None None
owner__richzeng richzeng
cc__chris@... cc__wsanchez
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_2"></a>|@ralphm set owner to chris-|
|-|-|

After quite some recent research into `twisted.python.log` and Python's `logging`, I am unconvinced about using `extra` as the way to pass additional data along so it can be used by `Formatter`s and/or `Handler`s. Internally, everything from `extra` is bolted onto the `LogRecord` instance as additional attributes. This means you have to be careful about already existing keys. E.g. it will prohibit overriding `message` and `asctime`. However it **doesn't** provide any protection against overriding methods. Damaging example:

```
>>> import logging
>>> logging.basicConfig()
>>> logging.warning('hello', extra={'getMessage': 'something'})
Traceback (most recent call last):
  File "logging/__init__.py", line 846, in emit
    msg = self.format(record)
  File "logging/__init__.py", line 723, in format
    return fmt.format(record)
  File "logging/__init__.py", line 464, in format
    record.message = record.getMessage()
TypeError: 'str' object is not callable
Logged from file <stdin>, line 1
```

Oops!

I'll explain a bit about Python's logging facilities below given the signature of the generic `Logging.log` method:

```python
    def log(self, level, msg, *args, **kwargs):
        """
        Log 'msg % args' with the integer severity 'level'.
        """
```

There are two different steps for formatting a log message through `Formatter.format` with the argument `record`, a `LogRecord` instance`.

 1. The record's `getMessage` is called. This basically does {{{msg = msg % record.args}}}, where `record.args` is usually what is passed in using the `args` in the above signature. But see below. The result is stored in `record.message`.
 2. The actual logged string is formatted. This is {{{s = self._fmt % record.__dict__}}}, where `s` is what is returned by `format`. If `record.exc_info` is set, a traceback will be rendered and added to `s`, too. The traceback is cached in `record.exc_text`.

If `args` has only item and the item is a dict and not empty, it is used instead (`args = args[0]`). This allows you to do this:

{{{
#!python
logging.log('WARNING', 'Hello, %{place}s!', {'place': 'World'})
}}}

In this usage, `record.args` will hold `{'place': 'World'}`, and `record.getMessage` yields ''Hello, World!''.

Depending how you want to actually the additional data from `eventDict` determines how we should pass it. If you need to rely on the standard `Formatter` behaviour and want to refer to keys from `eventDict` in the pre-set format string, you probably should use `extra`. However, if you need to write your own formatter and/or handler, you have direct access to the full `LogRecord` instance, including `args`, so that you don't get bitten by key clashes.

On top of all that, you probably also want to treat the `failure` key in `eventDict` if `isError` is set, so that the embedded exception information is passed on to the log record, too. You can pass this along as a tuple with the `exc_info` keyword argument to `Logging.log`.

As for the patch itself, it lacks a [ReviewProcess#Newsfiles news file] and we should probably document the added behaviour in our prose docs.

**Comment by twisted-trac** [comment]
|<img alt="chris-'s avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|chris- removed owner|
|-|-|

Replying to [ralphm](#note_2):
> After quite some recent research into `twisted.python.log` and Python's `logging`, I am unconvinced about using `extra` as the way to pass additional data along so it can be used by `Formatter`s and/or `Handler`s. Internally, everything from `extra` is bolted onto the `LogRecord` instance as additional attributes. This means you have to be careful about already existing keys. E.g. it will prohibit overriding `message` and `asctime`. However it **doesn't** provide any protection against overriding methods.

This is certainly true, but in my opinion, this is a problem that needs to be addressed in sdtlib logging, not by twisted.

> Depending how you want to actually the additional data from `eventDict` determines how we should pass it. If you need to rely on the standard `Formatter` behaviour and want to refer to keys from `eventDict` in the pre-set format string, you probably should use `extra`. However, if you need to write your own formatter and/or handler, you have direct access to the full `LogRecord` instance, including `args`, so that you don't get bitten by key clashes.

This is true, but the PythonLogObserver should not rely on the Handler and Formatter to handle this special case. The way I see it the extra keyword is the recommended way to pass structured data to the LogRecord and as stated above, it should not be up to twisted to work around problems the stdlib logging module.

> On top of all that, you probably also want to treat the `failure` key in `eventDict` if `isError` is set, so that the embedded exception information is passed on to the log record, too. You can pass this along as a tuple with the `exc_info` keyword argument to `Logging.log`.

Agreed, but when writing the patch, I did not consider this to be in the scope of the issue, especially since twisted already handles the traceback formatting for failures and exceptions. Passing the exc_info on as separate information would probably lead to logging each traceback twice.

> As for the patch itself, it lacks a [ReviewProcess#Newsfiles news file] and we should probably document the added behaviour in our prose docs.

Attached patch v2 adds a topfile and a mention of the behaviour in the api docs.

**Comment by twisted-trac** [comment]
|<img alt="richzeng's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|richzeng commented|
|-|-|

Please note that the current patch only deletes the 'message' key from the eventDict. There are several other keywords that Python stdlib's LogRecord keeps track of, and it complains when you try to pass them in using the 'extra' keyword argument.

They are documented here.
http://docs.python.org/library/logging.html#logrecord-attributes

**Comment by twisted-trac** [comment]
|<img alt="chris-'s avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|chris- commented|
|-|-|

Replying to [richzeng](#note_4):
> Please note that the current patch only deletes the 'message' key from the eventDict. There are several other keywords that Python stdlib's LogRecord keeps track of, and it complains when you try to pass them in using the 'extra' keyword argument.
> 
> They are documented here.
> http://docs.python.org/library/logging.html#logrecord-attributes

 1. The 'message' key is the only key that is mandatory to twisted log and incompatible to twisted, thus twisted should take care of cleaning that one up. This is not much of a problem because the message is passed to logging using a separate argument, so no data is lost here.

 2. You assume that sdtlib logging is done with the root logger. Because of that, you do not handle attributes of data that conflicts with possible LogRecord subclasses, so those will still fail. This applies to the logger and Logger subclasses as well.

Bottom line, you are trying to take filtering that normally for good reason takes place at the end of the log record creation chain and put it to the top. IMO that is a bad idea because it changes the way stdlib logging is meant to be used. Twisted should not try to predict what is meant to be done with data it passes to non-twisted libs.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

Reviewing.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

(In [34772]) Branching to 'structured-log-logging-5696'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/structured-log-logging-5696)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph set owner to richzeng<br>@glyph set status to `new`|
|-|-|

Thanks Rich,

  1. The approach taken to quash the problematic attributes in this patch has a number of problems.
    1. There's no tests for it.  What if this list of names were totally different?  Would the tests still pass?
    1. It's silently dropping data.  Don't do that.  The names of conflicting attributes should be mangled and then passed through.
    1. This last one is a silly nit-pick that I feel bad even making because I don't have benchmarks to back it up, but the list is re-created every time here rather than being stored in a global constant.  This means that you do an extra heap allocation.  It should be a `set()`.
  1. Something weird happened when I applied the patch, I guess.  It looks like `logger.log` gets invoked twice now.  Please do another patch from trunk to avoid this (hopefully it wasn't in your code!).  Tellingly, this _does not_ cause any tests to fail, which is a clear deficiency in the test suite.
  1. Twisted's coding standard is a tiny bit different from PEP8; please put _two_ blank lines between methods, _three_ between top-level suites (i.e. classes and module-level functions).
  1. Similarly I don't think the monkeying with whitespace elsewhere in the file is necessary (although maybe this is more patch-application weirdness).
  1. When the docstring refers to the `extra` parameter, it should use formatting to indicate that it's talking about a variable.  `C{}` is the epytext markup for fixed-width code text.

Sorry for all the back-and-forth on this, stdlib logging is pretty terrible (and ours is only slightly less bad).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

I'm working on this a bit in the context of [#6750](https://github.com/twisted/twisted/issues/6750).

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_11"></a>|@wsanchez commented|
|-|-|

[#6750](https://github.com/twisted/twisted/issues/6750) isn't going to do everything here.  See also its follow-up bug [#7167](https://github.com/twisted/twisted/issues/7167).

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_12"></a>|@wsanchez commented|
|-|-|

[#6750](https://github.com/twisted/twisted/issues/6750) landed, and actually it may address your needs here.
